packages/test/test-end-to-end-tests/src/test/incrementalSummaryContext.spec.ts

```diff
@@ -0,0 +1,617 @@
+				);
+			} else {
+				builder.addBlob(blobName, JSON.stringify(blobContent));
+			}
+		}
+
+		const content: ISnapshot = {
+			blobs: Array.from(this.blobMap.keys()),
+		};
+
+		builder.addBlob(headerBlobName, JSON.stringify(content));
+		return builder.getSummaryTree();
+	}
+	protected async loadCore(storage: IChannelStorageService): Promise<void> {
+		const content = await readAndParse<ISnapshot>(storage, headerBlobName);
+		for (const blob of content.blobs) {
+			const blobContent = await readAndParse<IBlob>(storage, blob);
+			this.blobMap.set(blob, blobContent);
+		}
+	}
+	protected processCore(
+		message: ISequencedDocumentMessage,
+		local: boolean,
+		localOpMetadata: unknown,
+	) {
+		if (message.type === MessageType.Operation) {
+			const op = message.contents as ICreateBlobOp;
+			switch (op.type) {
+				case "blobStorage": {
+					const blob: IBlob = {
+						value: op.value,
+						seqNumber: message.sequenceNumber,
+					};
+					const blobName = `${this.blobMap.size}`;
+					this.blobMap.set(blobName, blob);
+					break;
+				}
+				default:
+					throw new Error("Unknown operation");
+			}
+		}
+	}
+
+	public createBlobOp(content: string) {
+		const op: ICreateBlobOp = {
+			type: "blobStorage",
+			value: content,
+		};
+		this.submitLocalMessage(op);
+	}
+
+	protected onDisconnect() {}
+	protected applyStashedOp(content: any): unknown {
+		throw new Error("Method not implemented.");
+	}
+}
+
+// Test DDS factory for the tree dds
+class TestTreeDDSFactory implements IChannelFactory {
+	public static readonly Type = "incrementalTreeDDS";
+
+	public static readonly Attributes: IChannelAttributes = {
+		type: TestTreeDDSFactory.Type,
+		snapshotFormatVersion: "0.1",
+		packageVersion: pkgVersion,
+	};
+
+	public get type(): string {
+		return TestTreeDDSFactory.Type;
+	}
+
+	public get attributes(): IChannelAttributes {
+		return TestTreeDDSFactory.Attributes;
+	}
+
+	public async load(
+		runtime: IFluidDataStoreRuntime,
+		id: string,
+		services: IChannelServices,
+		attributes: IChannelAttributes,
+	): Promise<TestIncrementalSummaryTreeDDS> {
+		const sharedObject = new TestIncrementalSummaryTreeDDS(
+			id,
+			runtime,
+			attributes,
+			"TestTreeDDS",
+		);
+		await sharedObject.load(services);
+		return sharedObject;
+	}
+
+	/**
+	 * {@inheritDoc @fluidframework/datastore-definitions#IChannelFactory.create}
+	 */
+	public create(document: IFluidDataStoreRuntime, id: string): TestIncrementalSummaryTreeDDS {
+		return new TestIncrementalSummaryTreeDDS(id, document, this.attributes, "TestTreeDDS");
+	}
+}
+
+interface ISerializableTreeNode {
+	children: string[];
+	name: string;
+	seqNumber: number;
+}
+
+interface ITreeNode {
+	children: ITreeNode[];
+	name: string;
+	seqNumber: number;
+}
+
+interface ICreateTreeNodeOp {
+	parentPath: string[];
+	name: string;
+	type: "treeOp";
+}
+
+const rootNodeName = "rootNode";
+
+// Creates trees that can be incrementally summarized
```

Comment:
It's not trivial to follow what this DDS does. A comment here explaining that at high level would be helpful.